File: 126. Word Ladder II.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {string[][]}
 */
var findLadders = function(beginWord, endWord, wordList) {

    var ladders = new Array();
    var map = new Map();
    var distance = new Map();

    wordList.push(beginWord);
    //wordList.push(endWord);

    bfs(map, distance, beginWord, wordList);

    var path = new Array();

    dfs(ladders, path, beginWord, endWord, distance, map);

    return ladders;

    function bfs(map, distance, beginWord, wordList) {
        var q = new Array();
        q.push(beginWord);
        distance.set(beginWord, 0);
        for (var n in wordList) {
            map.set(wordList[n], []);
        }

        while (q.length > 0) {
            var crt = q.shift();
            var nextList = expand(crt, wordList);

            for ( var m in nextList ) {
                map.get(nextList[m]).push(crt);
                if (!distance.has(nextList[m])) {
                    distance.set(nextList[m], distance.get(crt) + 1);
                    q.push(nextList[m]);
                }
            }
        }
    }

    function expand (crt, wordList) {
        var expansion = new Array();

        var letters = 'abcdefghijklmnopqrstuvwxyz';
        for (var i = 0; i < crt.length; i++) {
            for (var j = 0; j < letters.length; j++) {
                if(crt[i] != letters[j]) {
                    var expanded = crt.substring(0, i) + letters[j] + crt.substring(i+1);

                    if (wordList.indexOf(expanded) > -1) {
                        expansion.push(expanded);
                    }
                }
            }
        }
        return expansion;
    }

    function dfs (ladders, path, beginWord, crt, distance, map) {
        path.push(crt);
        if( crt == beginWord ) {
            path.reverse();
            ladders.push(path.slice());
            path.reverse();
        } else {
            var crtList = map.get(crt);
            for (var x in crtList) {
                if (distance.has(crtList[x]) && distance.get(crt) == distance.get(crtList[x]) + 1) {
                    dfs(ladders, path, beginWord, crtList[x], distance, map);
                }
            }
        }
        path.pop();
    }
};

File: 126. Word Ladder II.py
"""
Given two words (beginWord and endWord), 
and a dictionary's word list, find all shortest transformation sequence(s) from beginWord to endWord, such that:

Only one letter can be changed at a time
Each transformed word must exist in the word list. Note that beginWord is not a transformed word.
Note:

Return an empty list if there is no such transformation sequence.
All words have the same length.
All words contain only lowercase alphabetic characters.
You may assume no duplicates in the word list.
You may assume beginWord and endWord are non-empty and are not the same.
Example 1:

Input:
beginWord = "hit",
endWord = "cog",
wordList = ["hot","dot","dog","lot","log","cog"]

Output:
[
  ["hit","hot","dot","dog","cog"],
  ["hit","hot","lot","log","cog"]
]
Example 2:

Input:
beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log"]

Output: []

Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
"""
"""
会出现编码错误，在lintcode上可以运行
"""
class Solution(object):
    Graph = {}
    result = []
    visited = set()

    def findLadders(self, beginWord, endWord, wordList):
        if endWord not in wordList:
            return self.result
        self.Graph[beginWord] = set()
        self.Graph[endWord] = set()
        for word in wordList:
            self.Graph[word] = set()

        self.buildGraph(beginWord, wordList)
        self.buildGraph(endWord, wordList)
        for word in wordList:
            self.buildGraph(word, wordList)

        dists = self.bfs(endWord)
        self.dfs(beginWord, endWord, [], dists)
        return self.result

    def buildGraph(self, word, dict):
        for i in xrange(len(word)):
            l = word[:i]
            r = word[i + 1:]
            for c in 'abcdefghijklmnopqrstuvwxyz':
                if word[i] != c:
                    curWord = l + c + r
                    if curWord in dict:
                        self.Graph[word].add(curWord)
                        self.Graph[curWord].add(word)

    def bfs(self, end):
        queue = []
        dists = {}
        queue.append(end)
        dists[end] = 0
        self.visited.add(end)

        while queue:
            cur = queue.pop(0)
            for child in self.Graph[cur]:
                if child not in self.visited:
                    queue.append(child)
                    self.visited.add(child)
                    dists[child] = dists[cur] + 1
        return dists

    def dfs(self, start, end, tmp, dists):
        tmp.append(start)
        if start == end:
            self.result.append(tmp[:])
            return

        for child in self.Graph[start]:
            if dists[child] != dists[start] - 1:
                continue
            self.dfs(child, end, tmp, dists)
            tmp.pop()